test/e2e/package_repo_test.go

```diff
@@ -5,15 +5,83 @@ package e2e
 
 import (
 	"fmt"
+	"reflect"
 	"strings"
 	"testing"
 	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/installpackage/v1alpha1"
+	kcv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
 )
 
 // TODO: Right now the implementation of the package repo reconciler needs improvement
 // hopefully after which, these tests can be cleaned up to remove retries and time related
 // falkeyness
 
+func Test_PackageRepoStatus_Failing(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, logger}
+	name := "repo"
+
+	repoYaml := `apiVersion: install.package.carvel.dev/v1alpha1
+kind: PackageRepository
+metadata:
+  name: test-repo
+spec:
+  fetch:
+    imgpkgBundle:
+      image: k8slt/i-dont-exist`
```

Comment:
Hmm thats an interesting question. I am leaning towards leaving it as an e2e test because I would like to see the tool's stderr in the Useful error message. That said, I think maybe talking about what makes e2e vs unit tests could be helpful. I usually think of unit as not needing outside dependencies, which the second test in that file actually requires since it is shelling out to ytt.

Reply:
failure scenarios imho are great tests to have at e2e level. it's as common code path as success case.

Reply:
FWIW, the tool's stderr would still be captured in the `usefulErrorMessage` in a non e2e test. Not saying e2e tests should not test errors. I am more saying that this can be tested without needing someone to set up what is needed for e2e tests (i.e. a Kubernetes cluster).

For me personally, I have been thinking e2e tests are needed when it would be valuable to actually have deployment take place as part of the test. Maybe something we can hash out a bit further as far as what appropriate testing levels are for kapp-controller.

I would say let's just revisit this test later on as I think it can be run quicker with less dependencies in a non e2e test.

Reply:
the objective here isnt to specifically test that usefulErrorMessage is being set, but rather here is what the failing status looks like for the real failing package repo. keep in mind that we are trying to capture user facing behaviour of failing pkg repo -- if tomorrow we redo package repo entirely in a different way this test should not be affected at all. (this is different from a test like this: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/app/app_reconcile_test.go where it's more coupled to implementation)

> without needing someone to set up what is needed for e2e tests (i.e. a Kubernetes cluster).

app_reconcile_test.go test has its own dependencies which could be considered more bothersome than k8s. + we need to have k8s anyway for other tests.

> I would say let's just revisit this test later on

sure, we can have longer discussion about this some time.